plotly/tools.py

```diff
@@ -1477,6 +1477,149 @@ class FigureFactory(object):
+
+        For documentation regarding to the form of the output, see
+        https://plot.ly/python/reference/#mesh3d-colorscale
+        """
+        colorscale = []
+        diff = 1./(len(colors) - 1)
+
+        for j in range(len(colors)):
+            colorscale.append([j * diff, colors[j]])
+        return colorscale
+
+    @staticmethod
+    def create_2D_density(x, y, colorscale='Earth', ncontours=20,
+                          hist_color=(0, 0, 0.5), pt_color=(0, 0, 0.5),
+                          pt_size=2, height=600, width=600):
+        """
+        Returns figure for a 2D density plot
+
+        :param (list) x: x-axis data for plot generation
+        :param (list) y: y-axis data for plot generation
+        :param (str|tuple|list) colormap: either a plotly scale name, an rgb
+            or hex color, a color tuple or a list or tuple of colors. An rgb
+            color is of the form 'rgb(x, y, z)' where x, y, z belong to the
+            interval [0, 255] and a color tuple is a tuple of the form
+            (a, b, c) where a, b and c belong to [0, 1]. If colormap is a
+            list, it must contain the valid color types aforementioned as its
+            members.
+        :param (int) ncontours: the number of 2D contours to draw on the plot
+        :param (str) hist_color: the color of the plotted histograms
+        :param (str) pt_color: the color of the scatter points
+        :param (str) pt_size: the color of the scatter points
+        :param (float) height: the height of the chart
+        :param (float) width: the width of the chart
+
+        Example 1: Simple 2D Density Plot
+        ```
+        import plotly.plotly as py
+        from plotly.tools import FigureFactory as FF
+
+        # Make data points
+        t = np.linspace(-1,1.2,2000)
+        x = (t**3)+(0.3*np.random.randn(2000))
+        y = (t**6)+(0.3*np.random.randn(2000))
+
+        # Create a figure
+        fig = FF.create_2D_density(x, y)
+
+        # Plot the data
+        py.iplot(fig, filename='simple-2d-density')
+        ```
+
+        Example 2: Use Parameters
+        ```
+        import plotly.plotly as py
+        from plotly.tools import FigureFactory as FF
+
+        # Make data points
+        t = np.linspace(-1,1.2,2000)
+        x = (t**3)+(0.3*np.random.randn(2000))
+        y = (t**6)+(0.3*np.random.randn(2000))
+
+        # Create custom colorscale
+        colorscale = ['#7A4579', '#D56073', 'rgb(236,158,105)',
+                      (1, 1, 0.2), (0.98,0.98,0.98)]
+
+        # Create a figure
+        fig = FF.create_2D_density(
+            x, y, colorscale=colorscale, hist_color='rgb(255, 237, 222)', pt_size=3)
+
+        # Plot the data
+        py.iplot(fig, filename='use-parameters')
+        ```
+        """
+        from plotly.graph_objs import graph_objs
+        colorscale = FigureFactory._validate_colors(colorscale, 'rgb')
+        colorscale = FigureFactory._make_colorscale(colorscale)
```

Comment:
I don't get it... what if colorscale is a string? How does the `_make_colorscale` work?